From 1dd800d86f4ab82a2ac29b3cdfd36019e1d3f763 Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Wed, 05 Jul 2023 08:47:54 +0300
Subject: [PATCH] [SW-151130] gaudi2: ppboot: clear arc0 dccm bss segment

The bss segment(DCCM) of arc0 in ppboot is used
as stack for threads in arcpid and all the thread
stack regions in bss are initialized to 0xAA pattern
In prstn reset when bringing pcie link up accessing
this bss is causing both arc to stuck which inturn
causes arc1 to get stuck. This patch adds support
to clear bss segment of arc0.

The issue is seen only in prstn because in cold reset
entire DCCM is cleared and in other resets the pcie link
will be already up.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: Iac06ae065cedbbe8b81098ea85c012d5063de6d9
---

diff --git a/arch/gaudi2/arc/secondary.S b/arch/gaudi2/arc/secondary.S
index cef77bb..f63167a 100644
--- a/arch/gaudi2/arc/secondary.S
+++ b/arch/gaudi2/arc/secondary.S
@@ -30,5 +30,11 @@
 	lpnz    1f
 	ld.ab   r5,[r1,4]
 	st.ab   r5,[r2,4]
+	/* Clear BSS */
+	mov     r6, __bss_start
+	sub     r7, __bss_end, r6
+	lsr.f   lp_count, r7, 2
+	lpnz    1f
+	st.ab   0, [r6, 4]
 1:
 	bl	secondary_main
